[test_ring_joint_sdpa.py] | Blackhole sanity tests | blackhole-multi-card-fast-sanity-tests / sdpa nightly tests (QB2 only) [bh_quietbox_2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] Ring Joint SDPA accuracy failures on Blackhole (PCC/RMSE below threshold) in test_ring_joint_sdpa.py
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** Blackhole sanity tests
**Job:** blackhole-multi-card-fast-sanity-tests / sdpa nightly tests (QB2 only) [bh_quietbox_2]

### Regression timeline

- **Last passing commit:** `6975081700ec` — 2026-07-18T12:25:42Z — https://github.com/tenstorrent/tt-metal/actions/runs/29644298491
- **First failing commit:** `7361a7ac160c` — 2026-07-19T12:26:13Z — https://github.com/tenstorrent/tt-metal/actions/runs/29686978437/job/88194384985
- **Most recent failing commit:** `35a716ccced5` — 2026-07-19T16:24:08Z — https://github.com/tenstorrent/tt-metal/actions/runs/29694784063/job/88214390020

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/29694784063/job/88214390020
- https://github.com/tenstorrent/tt-metal/actions/runs/29686978437/job/88194384985

### Failing test path(s)

- `tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py`
  - `test_ring_joint_attention_chunked_nd_sharded_indexed_kv_cache_accuracy[fp32_acc]`
  - `test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_smoke-fp32_acc]`
  - `test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_unicast_smoke-fp32_acc]`
  - `test_ring_joint_attention_sdpa_accuracy[mla_100k-fp32_acc]`
  - `test_ring_joint_attention_sdpa_accuracy[mla_128k-fp32_acc]`
  - `test_ring_joint_attention_sdpa_chunked_accuracy[kimi50k-all-qk-chunk2560-fp32_acc]`

### Error excerpt (terminal failure from logs)

```
=========================== short test summary info ============================
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_chunked_nd_sharded_indexed_kv_cache_accuracy[fp32_acc] - Failed: Chunked prefill mla_indexed_nd_sharded_kv-q32-k32 PCC/RMSE failures (pcc_threshold=0.997, rmse_threshold=0.05): chunk 0 (logical_n=256): PCC=0.9054393078700699, RMSE=0.145191
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_smoke-fp32_acc] - AssertionError: Main output minimax3_gqa_smoke-q128-k512 RMSE 0.111403 exceeds threshold 0.05
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_unicast_smoke-fp32_acc] - AssertionError: Main output minimax3_gqa_unicast_smoke-q128-k256 RMSE 0.093925 exceeds threshold 0.05
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[mla_100k-fp32_acc] - AssertionError: Main output mla_100k-q160-k160 PCC 0.9779359634848438 below threshold 0.997
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[mla_128k-fp32_acc] - AssertionError: Main output mla_128k-q128-k128 PCC 0.9862458344321603 below threshold 0.997
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_chunked_accuracy[kimi50k-all-qk-chunk2560-fp32_acc] - Failed: Chunked prefill kimi50k-q32-k512 PCC/RMSE failures (pcc_threshold=0.994, rmse_threshold=None): chunk 0 (logical_n=2560): PCC=0.7025292717617952, RMSE=0.135779
============ 6 failed, 104 passed, 86 skipped in 862.93s (0:14:22) =============
```

### Reproduction steps

Run the following on a Blackhole multi-card machine (QB2):

```bash
python -m pytest tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py \
  -k "fp32_acc" \
  -v
```

Or target specific failing variants:
```bash
python -m pytest \
  "tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_chunked_nd_sharded_indexed_kv_cache_accuracy[fp32_acc]" \
  "tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_smoke-fp32_acc]" \
  "tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_unicast_smoke-fp32_acc]" \
  "tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[mla_100k-fp32_acc]" \
  "tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_accuracy[mla_128k-fp32_acc]" \
  "tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_sdpa_chunked_accuracy[kimi50k-all-qk-chunk2560-fp32_acc]" \
  -v
```

### Notes

- All 6 failures are accuracy regressions: PCC values fall below their thresholds and/or RMSE values exceed their thresholds. This strongly suggests a numerical correctness regression introduced between `6975081700ec` (last passing, 2026-07-18) and `7361a7ac160c` (first failing, 2026-07-19).
- The failures span multiple test configs: MLA indexed ND-sharded KV cache (chunked prefill), minimax3 GQA (standard and unicast), MLA 100k/128k, and kimi50k chunked. All are multi-card ring joint SDPA tests on Blackhole.
- No timeouts or hardware errors observed; failures are pure numerical accuracy assertions.

**Determinism check:**
- **Run 1** (`35a716ccced5`): `FAILED test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_smoke-fp32_acc] - AssertionError: Main output minimax3_gqa_smoke-q128-k512 RMSE 0.111403 exceeds threshold 0.05` (6 failed, 104 passed, 86 skipped in 862.93s)
- **Run 2** (`7361a7ac160c`): `FAILED test_ring_joint_attention_sdpa_accuracy[minimax3_gqa_smoke-fp32_acc] - AssertionError: Main output minimax3_gqa_smoke-q128-k512 RMSE 0.111403 exceeds threshold 0.05` (6 failed, 104 passed, 86 skipped in 896.02s)
- All 6 failing tests, error messages, and PCC/RMSE values are byte-for-byte identical across both runs.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: Blackhole sanity tests`
`Auto-triage-job-name: blackhole-multi-card-fast-sanity-tests / sdpa nightly tests (QB2 only) [bh_quietbox_2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/29689763945
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1784470328182269?thread_ts=1784469215.568829&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
+++ b/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
@@ -45,13 +45,11 @@ void left_shift_init();
 void less_than_equal_zero_init();
 void less_than_zero_init();
 void logical_not_unary_init();
-void lrelu_init();
 void mask_init();
 void not_equal_zero_init();
 void power_init();
 void prelu_init();
 void relu_max_init();
-void relu_min_init();
 void reshuffle_rows_init();
 void right_shift_init();
 void selu_init();
@@ -94,11 +92,6 @@ inline void typecast_init() {
 inline void unused_init() { math::reset_counters(p_setrwc::SET_ABD_F); }
 
 }  // namespace sfpu
-
-// Dependent-false helper so the delegate's else-branch static_assert only fires when an uncategorized bare op is
-// instantiated (a plain static_assert(false) in an if-constexpr else is ill-formed pre-C++23).
-template <SfpuType>
-inline constexpr bool _sfpu_bare_op_unhandled_ = false;
 
 // Bare init entry point: delegates per SFPU op to its self-contained sfpu::<op>_init().
 template <SfpuType sfpu_op>
@@ -166,8 +159,6 @@ inline void llk_math_eltwise_unary_sfpu_init() {
         sfpu::less_than_zero_init();
     } else if constexpr (sfpu_op == SfpuType::logical_not_unary) {
         sfpu::logical_not_unary_init();
-    } else if constexpr (sfpu_op == SfpuType::lrelu) {
-        sfpu::lrelu_init();
     } else if constexpr (sfpu_op == SfpuType::mask) {
         sfpu::mask_init();
     } else if constexpr (sfpu_op == SfpuType::negative) {
@@ -180,8 +171,6 @@ inline void llk_math_eltwise_unary_sfpu_init() {
         sfpu::prelu_init();
     } else if constexpr (sfpu_op == SfpuType::relu_max) {
         sfpu::relu_max_init();
-    } else if constexpr (sfpu_op == SfpuType::relu_min) {
-        sfpu::relu_min_init();
     } else if constexpr (sfpu_op == SfpuType::reshuffle_rows) {
         sfpu::reshuffle_rows_init();
     } else if constexpr (sfpu_op == SfpuType::right_shift) {
@@ -222,10 +211,11 @@ inline void llk_math_eltwise_unary_sfpu_init() {
         // preceding compute_kernel_hw_startup, so route it to the full generic init.
         _llk_math_eltwise_unary_sfpu_init_<SfpuType::exponential>();
     } else {
-        static_assert(
-            _sfpu_bare_op_unhandled_<sfpu_op>,
-            "Bare SFPU_UNARY_INIT(OP) used for an op with no sfpu::<op>_init(). Add its self-contained init and a "
-            "delegate branch in llk_math_eltwise_unary_sfpu_init.h.");
+        // Generic fallback (pre-restructuring behavior): ops without a self-contained sfpu::<op>_init()
+        // get the full generic unary SFPU init (config reg + ADDR_MOD_7 + op-specific ADDR_MOD_6 via
+        // eltwise_unary_sfpu_configure_addrmod<OP>, which has a default valid for any SfpuType + counter
+        // reset). Keeps the bare no-arg delegate instantiable across the whole SfpuType set.
+        _llk_math_eltwise_unary_sfpu_init_<sfpu_op>();
     }
 }
 

--- a/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
+++ b/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
@@ -11,8 +11,10 @@
 
 namespace ckernel {
 
-// Kernel-invariant SFPU init: the SFPU config register + invariant ADDR_MOD_7, run once per kernel by the metal
-// "full init" entry points (compute_kernel_hw_startup / init_sfpu / unary_op_init_common / binary_op_init_common).
+// Kernel-invariant SFPU init (SFPU config register + invariant ADDR_MOD_7). Retained for the standalone tt-llk
+// SFPU test harness, which bypasses the metal "full init" entry points and runs this itself. The metal compute
+// path no longer hoists this: each per-op init below is self-contained (#50381), running the invariant per-op
+// rather than once-per-kernel.
 inline void llk_math_sfpu_init_once() { _llk_math_eltwise_unary_sfpu_init_once_(); }
 
 namespace sfpu {
@@ -65,9 +67,9 @@ void unary_le_init();
 void unary_lt_init();
 void unary_ne_init();
 
-// Self-contained per-op inits for ops used via bare SFPU_UNARY_INIT(OP) (no callback). config_reg + ADDR_MOD_7
-// are handled once per kernel by llk_math_sfpu_init_once(); these program only the op's residual state
-// (op-specific ADDR_MOD_6 where needed + reset the RWC counters).
+// Residual per-op inits for ops used via bare SFPU_UNARY_INIT(OP) (no callback). config_reg + ADDR_MOD_7 are
+// run per-op by the bare delegate below (_llk_math_eltwise_unary_sfpu_init_once_()), so these program only the
+// op's residual state (op-specific ADDR_MOD_6 where needed + reset the RWC counters).
 inline void fill_init() { math::reset_counters(p_setrwc::SET_ABD_F); }
 
 inline void isfinite_init() { math::reset_counters(p_setrwc::SET_ABD_F); }
@@ -101,6 +103,11 @@ inline constexpr bool _sfpu_bare_op_unhandled_ = false;
 // Bare init entry point: delegates per SFPU op to its self-contained sfpu::<op>_init().
 template <SfpuType sfpu_op>
 inline void llk_math_eltwise_unary_sfpu_init() {
+    // Per-op common SFPU init (config reg + invariant ADDR_MOD_7), formerly hoisted once-per-kernel via
+    // llk_math_sfpu_init_once(). Consolidated back per-op (#50381) so each init is fully self-contained and
+    // never depends on a separate once-init having run first. The co-located sfpu::<op>_init() below then
+    // programs only the op's residual state (op-specific ADDR_MOD_6 where needed + counter reset).
+    _llk_math_eltwise_unary_sfpu_init_once_();
     if constexpr (sfpu_op == SfpuType::abs) {
         sfpu::abs_init();
     } else if constexpr (sfpu_op == SfpuType::acos) {
@@ -222,10 +229,14 @@ inline void llk_math_eltwise_unary_sfpu_init() {
     }
 }
 
-// Callback init entry point (SFPU_UNARY_INIT_FN / _FN_ARGS / two-arg SFPU_UNARY_INIT): the op-specific init_func
-// is itself self-contained (it programs its own ADDR_MOD_6 if needed + resets counters). No generic prefix.
+// Callback init entry point (SFPU_UNARY_INIT_FN / _FN_ARGS / two-arg SFPU_UNARY_INIT). The per-op common init
+// (config reg + ADDR_MOD_7 + counter reset) runs first, then the op-specific init_func. Consolidated per-op
+// (#50381) rather than hoisted: re-asserting the shared SFPU config/addrmod state on every init -- on both the
+// MATH and PACK threads -- keeps the exp(PACK)/fp32-reciprocal(MATH) shared macro/replay programming from
+// interleaving destructively, which was the #50381 fp32 SDPA accuracy regression.
 template <SfpuType sfpu_op, class F, class... ARGS>
 inline void llk_math_eltwise_unary_sfpu_init(F&& init_func, ARGS&&... args) {
+    _llk_math_eltwise_unary_sfpu_init_<sfpu_op>();
     init_func(std::forward<ARGS>(args)...);
 }
 

--- a/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
+++ b/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
@@ -45,13 +45,11 @@ void left_shift_init();
 void less_than_equal_zero_init();
 void less_than_zero_init();
 void logical_not_unary_init();
-void lrelu_init();
 void mask_init();
 void not_equal_zero_init();
 void power_init();
 void prelu_init();
 void relu_max_init();
-void relu_min_init();
 void reshuffle_rows_init();
 void right_shift_init();
 void selu_init();
@@ -94,11 +92,6 @@ inline void typecast_init() {
 inline void unused_init() { math::reset_counters(p_setrwc::SET_ABD_F); }
 
 }  // namespace sfpu
-
-// Dependent-false helper so the delegate's else-branch static_assert only fires when an uncategorized bare op is
-// instantiated (a plain static_assert(false) in an if-constexpr else is ill-formed pre-C++23).
-template <SfpuType>
-inline constexpr bool _sfpu_bare_op_unhandled_ = false;
 
 // Bare init entry point: delegates per SFPU op to its self-contained sfpu::<op>_init().
 template <SfpuType sfpu_op>
@@ -166,8 +159,6 @@ inline void llk_math_eltwise_unary_sfpu_init() {
         sfpu::less_than_zero_init();
     } else if constexpr (sfpu_op == SfpuType::logical_not_unary) {
         sfpu::logical_not_unary_init();
-    } else if constexpr (sfpu_op == SfpuType::lrelu) {
-        sfpu::lrelu_init();
     } else if constexpr (sfpu_op == SfpuType::mask) {
         sfpu::mask_init();
     } else if constexpr (sfpu_op == SfpuType::negative) {
@@ -180,8 +171,6 @@ inline void llk_math_eltwise_unary_sfpu_init() {
         sfpu::prelu_init();
     } else if constexpr (sfpu_op == SfpuType::relu_max) {
         sfpu::relu_max_init();
-    } else if constexpr (sfpu_op == SfpuType::relu_min) {
-        sfpu::relu_min_init();
     } else if constexpr (sfpu_op == SfpuType::reshuffle_rows) {
         sfpu::reshuffle_rows_init();
     } else if constexpr (sfpu_op == SfpuType::right_shift) {
@@ -222,10 +211,11 @@ inline void llk_math_eltwise_unary_sfpu_init() {
         // preceding compute_kernel_hw_startup, so route it to the full generic init.
         _llk_math_eltwise_unary_sfpu_init_<SfpuType::exponential>();
     } else {
-        static_assert(
-            _sfpu_bare_op_unhandled_<sfpu_op>,
-            "Bare SFPU_UNARY_INIT(OP) used for an op with no sfpu::<op>_init(). Add its self-contained init and a "
-            "delegate branch in llk_math_eltwise_unary_sfpu_init.h.");
+        // Generic fallback (pre-restructuring behavior): ops without a self-contained sfpu::<op>_init()
+        // get the full generic unary SFPU init (config reg + ADDR_MOD_7 + op-specific ADDR_MOD_6 via
+        // eltwise_unary_sfpu_configure_addrmod<OP>, which has a default valid for any SfpuType + counter
+        // reset). Keeps the bare no-arg delegate instantiable across the whole SfpuType set.
+        _llk_math_eltwise_unary_sfpu_init_<sfpu_op>();
     }
 }
 

--- a/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
+++ b/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_init.h
@@ -11,8 +11,10 @@
 
 namespace ckernel {
 
-// Kernel-invariant SFPU init: the SFPU config register + invariant ADDR_MOD_7, run once per kernel by the metal
-// "full init" entry points (compute_kernel_hw_startup / init_sfpu / unary_op_init_common / binary_op_init_common).
+// Kernel-invariant SFPU init (SFPU config register + invariant ADDR_MOD_7). Retained for the standalone tt-llk
+// SFPU test harness, which bypasses the metal "full init" entry points and runs this itself. The metal compute
+// path no longer hoists this: each per-op init below is self-contained (#50381), running the invariant per-op
+// rather than once-per-kernel.
 inline void llk_math_sfpu_init_once() { _llk_math_eltwise_unary_sfpu_init_once_(); }
 
 namespace sfpu {
@@ -65,9 +67,9 @@ void unary_le_init();
 void unary_lt_init();
 void unary_ne_init();
 
-// Self-contained per-op inits for ops used via bare SFPU_UNARY_INIT(OP) (no callback). config_reg + ADDR_MOD_7
-// are handled once per kernel by llk_math_sfpu_init_once(); these program only the op's residual state
-// (op-specific ADDR_MOD_6 where needed + reset the RWC counters).
+// Residual per-op inits for ops used via bare SFPU_UNARY_INIT(OP) (no callback). config_reg + ADDR_MOD_7 are
+// run per-op by the bare delegate below (_llk_math_eltwise_unary_sfpu_init_once_()), so these program only the
+// op's residual state (op-specific ADDR_MOD_6 where needed + reset the RWC counters).
 inline void fill_init() { math::reset_counters(p_setrwc::SET_ABD_F); }
 
 inline void isfinite_init() { math::reset_counters(p_setrwc::SET_ABD_F); }
@@ -101,6 +103,11 @@ inline constexpr bool _sfpu_bare_op_unhandled_ = false;
 // Bare init entry point: delegates per SFPU op to its self-contained sfpu::<op>_init().
 template <SfpuType sfpu_op>
 inline void llk_math_eltwise_unary_sfpu_init() {
+    // Per-op common SFPU init (config reg + invariant ADDR_MOD_7), formerly hoisted once-per-kernel via
+    // llk_math_sfpu_init_once(). Consolidated back per-op (#50381) so each init is fully self-contained and
+    // never depends on a separate once-init having run first. The co-located sfpu::<op>_init() below then
+    // programs only the op's residual state (op-specific ADDR_MOD_6 where needed + counter reset).
+    _llk_math_eltwise_unary_sfpu_init_once_();
     if constexpr (sfpu_op == SfpuType::abs) {
         sfpu::abs_init();
     } else if constexpr (sfpu_op == SfpuType::acos) {
@@ -222,10 +229,14 @@ inline void llk_math_eltwise_unary_sfpu_init() {
     }
 }
 
-// Callback init entry point (SFPU_UNARY_INIT_FN / _FN_ARGS / two-arg SFPU_UNARY_INIT): the op-specific init_func
-// is itself self-contained (it programs its own ADDR_MOD_6 if needed + resets counters). No generic prefix.
+// Callback init entry point (SFPU_UNARY_INIT_FN / _FN_ARGS / two-arg SFPU_UNARY_INIT). The per-op common init
+// (config reg + ADDR_MOD_7 + counter reset) runs first, then the op-specific init_func. Consolidated per-op
+// (#50381) rather than hoisted: re-asserting the shared SFPU config/addrmod state on every init -- on both the
+// MATH and PACK threads -- keeps the exp(PACK)/fp32-reciprocal(MATH) shared macro/replay programming from
+// interleaving destructively, which was the #50381 fp32 SDPA accuracy regression.
 template <SfpuType sfpu_op, class F, class... ARGS>
 inline void llk_math_eltwise_unary_sfpu_init(F&& init_func, ARGS&&... args) {
+    _llk_math_eltwise_unary_sfpu_init_<sfpu_op>();
     init_func(std::forward<ARGS>(args)...);
 }
 

--- a/tt_metal/hw/inc/api/compute/compute_kernel_hw_startup.h
+++ b/tt_metal/hw/inc/api/compute/compute_kernel_hw_startup.h
@@ -86,9 +86,6 @@ ALWI void compute_kernel_hw_startup(uint32_t icb0, uint32_t icb1, uint32_t ocb) 
 
     MATH((llk_math_pack_sync_init<DST_ACCUM_MODE>()));
     MATH((llk_math_hw_configure<DST_ACCUM_MODE>(src_a_cb, src_b_cb)));
-    // Once-per-kernel SFPU init (SFPU config reg + invariant ADDR_MOD_7). Hoisted out of the per-op SFPU init
-    // so self-contained per-op inits (ckernel::sfpu::_init_<op>_) don't re-run it. Safe for non-SFPU kernels.
-    MATH((llk_math_sfpu_init_once()));
 
     PACK((llk_pack_hw_configure<DST_ACCUM_MODE>(ocb)));
     PACK((llk_pack_init<PackMode::Default>(ocb)));

--- a/tt_metal/hw/inc/api/compute/eltwise_binary.h
+++ b/tt_metal/hw/inc/api/compute/eltwise_binary.h
@@ -38,11 +38,6 @@ ALWI void binary_op_init_common(uint32_t icb0, uint32_t icb1, uint32_t ocb, uint
 
     MATH((llk_math_pack_sync_init<DST_ACCUM_MODE>()));
     MATH((llk_math_hw_configure<DST_ACCUM_MODE>(icb0, icb1)));
-    // Once-per-kernel SFPU init (SFPU config reg + invariant ADDR_MOD_7). SFPU ops are frequently used after
-    // binary_op_init_common (softmax, moreh_*), so this full-init entry point must also carry the hoisted
-    // once-init that self-contained per-op inits (ckernel::sfpu::_init_<op>_) rely on. Idempotent for pure
-    // binary kernels.
-    MATH((llk_math_sfpu_init_once()));
 
     PACK((llk_pack_hw_configure<DST_ACCUM_MODE>(ocb)));
     PACK((llk_pack_init(ocb)));

--- a/tt_metal/hw/inc/api/compute/eltwise_unary/eltwise_unary.h
+++ b/tt_metal/hw/inc/api/compute/eltwise_unary/eltwise_unary.h
@@ -35,10 +35,6 @@ ALWI void unary_op_init_common(uint32_t icb, uint32_t ocb, uint32_t call_line = 
     // Asserted after hw_configure (which sets the operand-driven DEFAULT) so it is the last writer
     // before the op runs; skip-if-set keeps it cheap.
     MATH((ckernel::math::_configure_unary_preserve_zero_flag_state_()));
-    // Once-per-kernel SFPU init (SFPU config reg + invariant ADDR_MOD_7). Hoisted out of the per-op SFPU init
-    // so self-contained per-op inits (ckernel::sfpu::_init_<op>_) don't re-run it. init_sfpu() delegates here,
-    // so it inherits this call.
-    MATH((llk_math_sfpu_init_once()));
 #else
     UNPACK((llk_unpack_hw_configure(icb)));
     UNPACK((llk_unpack_A_init<BroadcastType::NONE, false, EltwiseBinaryReuseDestType::NONE, UnpackToDestEn>(

